core/src/main/java/org/infinispan/container/impl/SegmentedDataContainer.java

```diff
@@ -0,0 +1,169 @@
+package org.infinispan.container.impl;
+
+import java.util.Iterator;
+import java.util.Spliterator;
+import java.util.function.Consumer;
+
+import org.infinispan.commons.util.IntSet;
+import org.infinispan.container.DataContainer;
+import org.infinispan.container.entries.InternalCacheEntry;
+import org.infinispan.metadata.Metadata;
+
+/**
+ * Interface describing methods of a data container where operations can be indexed by the segment of the key
+ * stored in the map. This allows for much more efficient iteration when only a subset of segments are required for
+ * a given operation (which is the case very often with Distributed caches).
+ * @author wburns
+ * @since 9.3
+ */
+public interface SegmentedDataContainer<K, V> extends DataContainer<K, V> {
```

Comment:
I would have liked to merge `SegmentedDataContainer` into `DataContainer` in this PR, not later

Reply:
So I thought we wanted to separate the interfaces? Or now you want all the methods on `DataContainer`?